marshalers.go

```diff
@@ -85,37 +85,30 @@ func marshalPerCPUValue(slice any, elemLength int) (sys.Pointer, error) {
 // possible CPU.
 //
 // slicePtr must be a pointer to a slice.
-func unmarshalPerCPUValue(slicePtr any, elemLength int, buf []byte) error {
-	slicePtrType := reflect.TypeOf(slicePtr)
-	if slicePtrType.Kind() != reflect.Ptr || slicePtrType.Elem().Kind() != reflect.Slice {
-		return fmt.Errorf("per-cpu value requires pointer to slice")
+func unmarshalPerCPUValue(slice any, elemLength int, buf []byte) error {
+	sliceType := reflect.TypeOf(slice)
+	if sliceType.Kind() != reflect.Slice {
+		return fmt.Errorf("per-cpu value requires a slice")
 	}
 
 	possibleCPUs, err := internal.PossibleCPUs()
 	if err != nil {
 		return err
 	}
 
-	sliceType := slicePtrType.Elem()
-	slice := reflect.MakeSlice(sliceType, possibleCPUs, possibleCPUs)
+	sliceValue := reflect.ValueOf(slice)
 
 	sliceElemType := sliceType.Elem()
 	sliceElemIsPointer := sliceElemType.Kind() == reflect.Ptr
-	if sliceElemIsPointer {
-		sliceElemType = sliceElemType.Elem()
-	}
-
 	stride := internal.Align(elemLength, 8)
 	for i := 0; i < possibleCPUs; i++ {
 		var elem any
+		v := sliceValue.Index(i)
```

Comment:
This can panic since there is no bounds check against `sliceValue.Length()`. I think it makes sense to move the check for length from ensurePerCPUSlice into this function.

Reply:
I think I should copy the check in, though it'll mean it's impossible to test against the public API, since the validation happens at the outer level.

Reply:
Added a test against the inner API too.

Reply:
I pushed eea7482ab581845bc2f74a6e0b753e186d3ff5c0 which implements what I meant: all of the reflection specific validation happens in `unmarshalPerCPUValue`. PTAL!

Reply:
oh I see you meant to inline validation as you make progress (but before the syscall), vs upfront validation. Yeah technically there is no harm here since you aren't actually doing anything.

Also I forgot about `github.com/frankban/quicktest`.